tests/spec_decode/test_multi_step_worker.py

```diff
@@ -440,3 +564,116 @@ def test_draft_proposals_mixed_k():
+    expanded_request, indices_of_original_sequence_groups =\
+        MultiStepWorker._expand_execute_model_request(
+            execute_model_request,
+            set(seq_ids_with_bonus_tokens))
+    # Validate that the number of sequence groups is now the original number
+    # plus the number of sequences with bonus tokens.
+    assert len(expanded_request.seq_group_metadata_list) == \
+        len(prompts) + len(seq_ids_with_bonus_tokens)
+    # Iterate through the updated request and validate the following:
+    # 1. If the sequence group is part of the original request, it contains
+    #    all token ids including the bonus tokens.
+    # 2. If the sequence group is newly added, it does not contain the
+    #    bonus token.
+    for index, seq_group_metadata in enumerate(
+            expanded_request.seq_group_metadata_list):
+        seq_id = next(iter(seq_group_metadata.seq_data.keys()))
+        assert seq_group_metadata.seq_data[seq_id].prompt_token_ids ==\
+            seq_id_prompt_map[seq_id]
+        if index in indices_of_original_sequence_groups:
+            assert seq_group_metadata.seq_data[seq_id].output_token_ids ==\
+                seq_id_output_token_map[seq_id]
+        else:
+            assert seq_group_metadata.seq_data[seq_id].output_token_ids ==\
+                seq_id_output_token_map[seq_id][:-1]
+
+
+@torch.inference_mode()
+@pytest.mark.parametrize('num_steps', [1, 2, 6])
+@pytest.mark.parametrize('batch_size', [1, 32, 64])
+def test_filter_model_output(num_steps: int, batch_size: int):
+    """
+    Test the _filter_model_output function of the MultiStepWorker class.
+
+    This test ensures that the _filter_model_output method correctly filters the
+    model's output, retaining only the specified sequences.
+    """
+    vocab_size = 32_000
+
+    target_token_ids = torch.randint(low=0,
+                                     high=vocab_size,
+                                     size=(batch_size, (num_steps)),
+                                     dtype=torch.int64,
+                                     device='cuda')
+    target_token_probs = torch.rand(batch_size,
+                                    num_steps,
+                                    vocab_size,
+                                    dtype=torch.float32,
+                                    device='cuda')
+    target_token_logprobs = torch.rand(batch_size,
+                                       num_steps,
+                                       vocab_size,
+                                       dtype=torch.float32,
+                                       device='cuda')
+    sampler_output_list = create_sampler_output_list(target_token_ids,
+                                                     target_token_probs,
+                                                     target_token_logprobs)
+    output_indices_to_retain = random.sample(range(num_steps),
+                                             max(1, num_steps // 2))
```

Comment:
I am confused why we sample from `range(num_steps)`. shouldn't it be `range(batch_size)`?

Reply:
It should be range(batch_size). I removed the test though in favor of test_same_output_for_multi_step_with_batch_expansion